### PR TITLE
Basic setup for `TableBlock`

### PR DESCRIPTION
**Pull request by jamesricky**
## Description

This is the basic API/Admin setup for the `TableBlock`. 

A bunch of required features are still missing so we can have multiple small PRs instead of a single huge/complicated one. 
Things that will be added in future PRs into the feature-branch, assuming this PR will be approved in it's current form:
- Changeset
- Complete all of the `TODO` code comments ([#2808](https://github.com/vivid-planet/comet/pull/2808), [#2805](https://github.com/vivid-planet/comet/pull/2805))
- Adding/Removing/Reordering of columns and rows ([#2808](https://github.com/vivid-planet/comet/pull/2808))
- Set style-settings, such as highlighting/size to columns and rows ([#2808](https://github.com/vivid-planet/comet/pull/2808))
- Rich-text editing of cell-content

## Screenshots/screencasts

Table Grid

| Editing single line text | Editing multi line text |
| --- | --- |
| Editing single line text | Editing multi line text |

## Related tasks and documents

https://vivid-planet.atlassian.net/browse/COM-1015

## Open TODOs/questions

- [x] Is it a problem, that we add `@mui/x-data-grid-pro` to `@comet/cms-admin`? 
Features, such as row- and column-reordering are a pro feature. 
Do we maybe need a separate package for the `TableBlock` to prevent requiring `@mui/x-data-grid-pro` in all applications that use `@comet/cms-admin`?

### CODE REVIEW

File: demo/site/src/blocks/TableBlock.tsx [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { PropsWithData } from "@comet/cms-site";
+import { TableBlockData } from "@src/blocks.generated";
+
+// TODO: Used for debugging - either remove this or implement properly before merging feature branch
```
Comment (johnnyomair) [inline demo/site/src/blocks/TableBlock.tsx:4]: Yes please

File: packages/api/cms-api/src/blocks/table.block.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @BlockField()
+    position: number;
+
+    @BlockField()
+    highlighted: boolean;
+
+    @BlockField(TableBlockRowColumnValueData)
+    @Type(() => TableBlockRowColumnValueData)
+    columnValues: TableBlockRowColumnValueData[] = [];
```
Comment (johnnyomair) [inline packages/api/cms-api/src/blocks/table.block.ts:83]: Wouldn't cells be a better name?

File: packages/api/cms-api/src/blocks/table.block.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+class TableBlockColumnData extends BlockData {
+    @BlockField()
+    id: string;
+
+    @BlockField()
+    position: number;
+
+    @BlockField({ type: "enum", enum: ColumnSize })
+    size: ColumnSize;
```
Comment (johnnyomair) [inline packages/api/cms-api/src/blocks/table.block.ts:21]: Should this be configurable?

**Review by johnnyomair — COMMENTED** [review]
Don't we need support for rich text content in the table block? Should the cell content really be only a string? Wouldn't it be more flexible if it would accept a block?

**Comment by johnnyomair** [comment]
> Is it a problem, that we add @mui/x-data-grid-pro to @comet/cms-admin?
Features, such as row- and column-reordering are a pro feature.

We previously discussed to somehow add support for all data grid versions in the library (maybe a context). IMO it would be nice if the library would work without the pro version. Maybe custom context menu items for moving rows/columns would work. But this is out-of-scope

> Do we maybe need a separate package for the TableBlock to prevent requiring @mui/x-data-grid-pro in all applications that use @comet/cms-admin?

I wouldn't add a separate package for this because it increases the required maintenance efforts.

@thomasdax98 @nsams what do you think?

**Comment by jamesricky** [comment]
> Don't we need support for rich text content in the table block? Should the cell content really be only a string? Wouldn't it be more flexible if it would accept a block?

I'd leave it as a string for now, allowing rich-text content will be a follow-up PR.

File: packages/api/cms-api/src/blocks/table.block.ts (hunk 3 above)
Comment (jamesricky) [reply packages/api/cms-api/src/blocks/table.block.ts:21]: Do you mean the values or the setting itself? 
I think the 5 values already defined should be sufficient.

File: demo/site/src/blocks/TableBlock.tsx (hunk 1 above)
Comment (jamesricky) [reply demo/site/src/blocks/TableBlock.tsx:4]: Will be done in a follow-up PR, I'll keep this for now, as it's helpful during development.

File: packages/api/cms-api/src/blocks/table.block.ts (hunk 3 above)
Comment (johnnyomair) [reply packages/api/cms-api/src/blocks/table.block.ts:21]: The setting itself. As an editor I'd except that the table would fit the content automatically.

**Comment by jamesricky** [comment]
> What about other blocks like images or link lists? Also I suppose we need links in the cells, so using a rich text block is inevitable IMO. Maybe we should discuss this with UX.

Yes, rich text is a requirement for things like inline-formatting, lists and links. 
UX already created a design for this:

Screenshot 2024-11-12 at 14 20 54

---

We haven't considered images yet, I suppose this is something we should discuss.

File: packages/api/cms-api/src/blocks/table.block.ts (hunk 3 above)
Comment (jamesricky) [reply packages/api/cms-api/src/blocks/table.block.ts:21]: > As an editor I'd except that the table would fit the content automatically.

Yes, it should and will. 

I don't think showing an optional setting in the admin hurts but we would always have to implement this behavior in the site block then 🤔 

Will discuss this with UX, if we ever explicitly don't want admins to be able to control the column size.

File: packages/api/cms-api/src/blocks/table.block.ts (hunk 2 above)
Comment (jamesricky) [reply packages/api/cms-api/src/blocks/table.block.ts]: Done in 2b248cb05e0da069892384e241914b6850cbebd9

File: packages/api/cms-api/src/blocks/table.block.ts (hunk 3 above)
Comment (jamesricky) [reply packages/api/cms-api/src/blocks/table.block.ts:21]: As discussed with UX, there is no need to make this configurable.

File: packages/admin/cms-admin/package.json [hunk 4]
```diff
@@ -151,6 +151,7 @@
         "@emotion/styled": "^11.3.0",
         "@mui/material": "^5.0.0",
         "@mui/x-data-grid": "^5.0.0",
+        "@mui/x-data-grid-pro": "^5.0.0",
```
Comment (johnnyomair) [inline packages/admin/cms-admin/package.json:154]: In the admin package we made this peer dependency optional, would this work here as well? https://github.com/vivid-planet/comet/blob/main/packages/admin/admin/package.json#L119-L124

Comment (jamesricky) [reply packages/admin/cms-admin/package.json:154]: Cool! I think so, added in 6ca0fc6cda80e37ca33c95d253fe2e505e52e539

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**